_posts/2015-06-25-infrastructure-as-code.md

```diff
@@ -0,0 +1,82 @@
+---
+layout: post
+title: "Infrastructure as code: automating Jenkins"
+date: "2015-06-25 09:00:00"
+icon: wrench
+tags: infrastructure, configuration management, continuous integration, jenkins, automation
+categories: tech-stories
+authors: ["Jens Fischer", "Dirk Jablonski"]
+---
+
+Here at ePages we heavily rely on [Continuous Integration (CI)](http://www.martinfowler.com/articles/continuousIntegration.html) to make sure our code is always in a good shape.
+[Jenkins](https://jenkins-ci.org/) is the software we use as our CI tool.
+A mixed language stack mainly composed of `Java`, `JavaScript` and `Perl` spread over multiple teams brings in different build processes and toolings needed to produce the artifacts our software is composed of, e.g. `gradle`, `grunt` and `make`.
+We do not run all build jobs on a single Jenkins instance, but already spread out different jobs to dedicated Jenkins servers.
+Nevertheless some teams still share their build infrastructure, thus lacking independence e.g. when it comes to introducing new Jenkins plugins or other tools needed for their build process.
+
+{% image blog-jenkins-polyglot.png %}{% endimage %}
+
+# Jenkins configuration
+
+Jenkins makes it very easy to introduce new functionality by installing one of the [many available plugins](https://wiki.jenkins-ci.org/display/JENKINS/Plugins).
+Unfortunately not every plugin turns out to work as expected or in the worst case can potentially bring down the whole Jenkins instance.
+To evaluate unknown plugins by installing them into your production CI infrastructure bears problems, since they often leave unwanted traces behind, even after uninstallation.
+Even making changes to a job configuration without adding any new plugins can lead to broken build jobs, negatively affecting the productivity of whole teams.
+
+## The solution
+
+We were looking for ways to make it safe and easy to improve our build infrastructure.
+Using version control like [GitHub](https://github.com/) for all of our source code, it was natural to treat our infrastructure the same way and be able to roll back to a previous version without any hassle.
+Versioning whole images of Jenkins virtual servers didn't seem feasible and plugins like [JobConfigHistory](https://wiki.jenkins-ci.org/display/JENKINS/JobConfigHistory+Plugin) didn't go far enough.
```

Comment:
"didn't go far enough" is a little unclear. It didn't go where? or do what?
I would suggest

"plugins like xx were not comprehensive enough"
or "plugins like xx were not thorough enough"
or "plugins like xx did not meet our requirements"